api/endpoints/course.py

```diff
@@ -305,3 +308,100 @@ async def buy_course(user: User = user_auth, course: Course = get_course) -> Any
     await clear_cache("course_access")
 
     return True
+
+
+@router.get("/courses_viewtime", responses=responses(ViewTime))
+async def get_course_viewtime(user: User = user_auth) -> Any:
+    """
+    Return the total viewtime of all courses.
+
+    *Requirements:* **VERIFIED**
+    """
+
+    completed_lectures: dict[str, set[str]] = {}
+
+    async for lecture in await db.stream(filter_by(models.LectureProgress, user_id=user.id)):
+        completed_lectures.setdefault(lecture.course_id, set()).add(lecture.lecture_id)
+
+    lecture_data = [
+        course.summary(None if completed_lectures is None else completed_lectures.get(course.id, set()))
+        for course in iter(COURSES.values())
+    ]
+
+    sub_skill_reponses = []
+
+    for sub_skill in lecture_data:
+        total_time = 0
+        sections = []
+
+        for section in sub_skill.sections:
+            section_time = 0
+            lectures = []
+
+            for lecture in section.lectures:
+                if lecture.duration > 0 and lecture.completed:
+                    lectures.append(ViewTimeLecture(lecture_name=lecture.title, time=lecture.duration))
+                    section_time += lecture.duration
+
+            if section_time > 0:
+                sections.append(ViewTimeSection(section_name=section.title, total_time=section_time, lectures=lectures))
+                total_time += section_time
+
+        if total_time > 0:
+            sub_skill_reponses.append(
+                ViewTimeSubSkill(
+                    sub_skill_id=sub_skill.id, sub_skill_name=sub_skill.title, total_time=total_time, sections=sections
+                )
+            )
+
+    return ViewTime(
+        total_time=sum([sub_skill.total_time for sub_skill in sub_skill_reponses]), sub_skills=sub_skill_reponses
+    )
+
+@router.get("/tasks_viewtime", responses=responses(TotalTime, DataFetchError, UnexpectedDataStructure))
+async def get_tasks_viewtime(user: User = user_auth) -> Any:
+    """
+    Return the total viewtime of all tasks.
+
+    *Requirements:* **VERIFIED**
+    """
+    url = "https://api.test.bootstrap.academy/challenges/subtasks"  # TODO: Replace by a real URL
+    # headers = {"Authorization": f"Bearer {user.auth_token}"}
+    # headers = {"Authorization": f"Bearer testBearerToken"}
+    headers = {"Authorization": f"Bearer eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE3MzAyMzU4MjYsInVpZCI6ImFlYTY4OWI1LWViYWYtNDg5Yy1iMDU0LTNiOTlhNDY1NTU2MSIsInNpZCI6IjI1ZDQ3NjM5LWRjNTAtNGZiYi05M2FkLWY4MmRiZTUyM2E1MCIsInJ0IjoiYmVjZTZhNTc5NjEwMjBiMjcyZDg2YTNmNjZjODc0MjM3NjgxNjUzY2MxYWI0YjU5OGZiMmFhNmU0YmY4NmIxYSIsImRhdGEiOnsiYWRtaW4iOnRydWUsImVtYWlsX3ZlcmlmaWVkIjp0cnVlfX0.8unbpeiD-oW652aVkOLvUMIpDsTDgY-kk84zXjUZSh8"}
```

Comment:
Please use the `InternalService` class to communicate with other microservices. There is an example for shop-ms [here](https://github.com/Bootstrap-Academy/skills-ms/blob/develop/api/services/shop.py).

Reply:
I have now implemented the `InternalService`. Should I now send a request to `/subtasks` or should I copy the code and send the request to `/_internal/subtasks`? I'm asking this because I actually want to avoid code duplication.

Reply:
Ideally this should be sent to a dedicated internal endpoint in the challenges microservice. However I think this microservice does not have any internal endpoints yet, so in this case, for a quick and dirty solution, you *could* just forward the user's access token to the public endpoint of challenges-ms.